### PR TITLE
JSON to proto.Message Unmarshaler

### PR DESCRIPTION
**Pull request by Codehardt**
Wrote a JSON to proto.Message that uses the `age`, `type`, `value` construct of json.

### CODE REVIEW

File: unmarshal.go [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+// unmarshalTypeValue unmarshals a value to a predefined object. The
+// value has to be a pointer to bool, int, float, string or struct.
+// If the value is a pointer to struct, the grrType is used to unmarshal
+// into proto.Message.
+func unmarshalTypeValue(obj interface{}, grrType string, grrValue interface{}) error {
+
+	protoType := reflect.TypeOf(obj)
+	protoValue := reflect.ValueOf(obj).Elem()
+
+	/*
```
Comment (hillu) [inline unmarshal.go]: Please use C++-style comments, don't use extra lines

File: unmarshal.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+import (
+	"encoding/json"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"reflect"
+	"strings"
+)
+
+// Unmarshal unmarshals a byte array into a predefined proto.Message.
+func Unmarshal(b []byte, pm proto.Message) error {
```
Comment (hillu) [inline unmarshal.go]: This function does not need to be public.

File: unmarshal.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+				av, err := NewAnyValue(newObj.(proto.Message))
+				if err != nil {
+					return err
+				}
+				protoValue.Set(reflect.ValueOf(av).Elem())
+			} else {
+				protoValue.Set(reflect.ValueOf(newObj).Elem())
+			}
+		}
+	case reflect.Slice:
```
Comment (hillu) [inline unmarshal.go]: If this case is not handled, it should not be there in the first place.

File: unmarshal.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+		}
+		protoValue.Set(reflect.ValueOf(unmarshaler).Elem())
+		return nil
+	}
+
+	/*
+	 * Else check what kind of interface the
+	 * object to be written is
+	 */
+	switch protoType.Elem().Kind() {
```
Comment (hillu) [inline unmarshal.go]: It should be possible to do this using a type switch instead of going through the `reflect` API

File: unmarshal.go [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+		return err
+	}
+	grrValue, hasValue := jsonMap["value"]
+	grrType, hasType := jsonMap["type"].(string)
+	if hasValue && hasType {
+		return unmarshalTypeValue(pm, grrType, grrValue)
+	} else {
+		return unmarshalMap(pm, jsonMap)
+	}
+	return nil
```
Comment (hillu) [inline unmarshal.go:26]: Which concrete types go through unmarshalTypeValue, which go through unmarshalMap?

File: apiclient.go [hunk 6]
```diff
@@ -90,7 +90,7 @@ func (c *APIClient) DoRequest(rq *http.Request) (*http.Response, error) {
 func (c *APIClient) do(method, apipath string, rqm, rsm proto.Message) error {
 	u := *c.BaseURL
 	u.Path = path.Join(u.Path, apipath)
-	u.RawQuery = "strip_type_info=1"
+	//u.RawQuery = "strip_type_info=1"
```
Comment (hillu) [inline apiclient.go]: Please don't comment out code like this.

File: unmarshal.go (hunk 5 above)
Comment (Codehardt) [reply unmarshal.go:26]: I can't provide a full list of types that go through the typeValue- and through the map-unmarshaler, it depends on how the GRR-Response is constructed.

In some cases, e.g. `/api/clients/C.1000000000000000/flows/F:ABCDEF12`, GRR returns the TypeValue construct in the top level:
`{ "age": 0, "type": "ApiFlow", "value": {...}}`

In other cases, e.g. `/api/clients/C.1000000000000000/last-ip`, GRR returns the object directly:
`{ "info": "Internal IP address.", "ip": "192.168.100.42", "status": "INTERNAL" }`

**Pull request closed without merging**